Dragndrop - add enhanced dragndrop lists

BurningTreeC opened the pull request:
This draft PR adds an enhanced `make-draggable-droppable` and `draggable-droppable-list` procedure:

https://github.com/user-attachments/assets/df7803e4-7b3c-44ad-8986-af218c0722ba

currently it works for vertically aligned lists only but it will be enhanced for horizontal lists, too
the transitions you see in the video can still be refined, I'm a bit tired now 😁 and will have to have a look with fresh eyes

I can imagine that this could for example easily be used for TiddlyDesktop's landing page ...

Please let me know if this is possible core material. Thank you <3

BurningTreeC commented:
missing: handling for incomplete drop

pmario commented on core/wiki/macros/dragndrop.tid:
```diff
@@ -0,0 +1,378 @@
+</$let>
+\end
+
+\function get.drag.state.tiddler.height() [<dragStateTiddler>get[height]addsuffix[px]]
```
There is already a convention for core global filter naming. They should be prefixed with `tf.` and named-with-hyphens.

In this case, all functions use a `get.` prefix, which imo is redundant, since functions always "get" a variable

So I thin for consistency reasons it should be named `tf.drag-state-tiddler-height`

pmario commented on core/wiki/macros/dragndrop.tid:
```diff
@@ -0,0 +1,378 @@
+\function get.tiddler.before.target() [<dragStateTiddler>get[target]] :reduce[list<list>before<currentTiddler>]
+\function get.tiddler.after.target() [<dragStateTiddler>get[target]] :reduce[list<list>after<currentTiddler>]
+
+\function get.styleTransition()
```
Also here `tf.styleTransition` should be enough since functions always "get"

pmario commented:
>This draft PR adds an enhanced make-draggable-droppable and draggable-droppable-list procedure:

Sorry for my ignorance, but I am not exactly sure, what the enhancement is.

pmario commented:
Also I think allocating the height of the tiddler in the story river does not make any sense for the default views available at tw-com. Is there an other usecase?

Jermolene commented:
Hi @BurningTreeC I like this. I’m interested in trying to evolve the UX closer to the standards of modern web applications, and this is a tour de force that nicely demonstrates how much of this sort of functionality can now be done in wiki text.

The core size increase is very significant, and it would be good to do all that we can to reduce it. Even if (probably optimistically) we got the size increase down to 15KB, I think it may still make sense to go ahead with it. The code is nicely encapsulated, so particularly if there’s a case that we can reuse a significant amount of it, I’d be interested to explore further.

BurningTreeC commented:
Hi @pmario and @Jermolene 
I've come back to this and made it work much better
Please have a look at the preview build and the options in `$:/ControlPanel` > `Appearance` > `Drag & Drop`

BurningTreeC commented:
https://github.com/user-attachments/assets/64e85736-c6bc-470b-9ecf-cac4d4e27b6e

https://github.com/user-attachments/assets/b80ee6bb-bf79-4a3a-abd5-886d47176afc

pmario commented on core/ui/SideBar/Open.tid:
```diff
@@ -5,33 +5,21 @@ caption: {{$:/language/SideBar/Open/Caption}}
 \end
 
-<div class="tc-sidebar-tab-open">
+<div class="tc-sidebar-tab-open" style.border-top-style="solid" style.border-top-width="1px" style.border-top-color="transparent">
```
IMO the need to hardcode styles like this seems to be a bug. It makes it impossible to style elements in the open tab with CSS rules. 
Also the "insertion placeholder" seems to be missing now, which for me is a regression

BurningTreeC replied:
Hi @pmario , I've removed these styles now. They were needed for a test before this, now not needed anymore.
Also, in the ControlPanel you can now choose a setting, which shows placeholders again

pmario commented:
IMO updating the story river while the elements are dragged in the Open sidebar should be avoided.

IMO It adds unnecessary distraction to the river, if the user eg: "overshoots" to the top and then drags the tiddler back down by one or 2 elements.

If there is animation at the story river while dragging the Open tab elements is distracting.

My measurement for animation always is: **Does it serve a purpose**. If something only is animated, because we can, it should not be animated at all.

If an animation makes the user loose context or loose focus it should be not animated. So if I drag and drop elements in the open sidebar and the story river animates, **I do loose focus**

BurningTreeC commented:
> IMO updating the story river while the elements are dragged in the Open sidebar should be avoided.
> 
It's now decoupled. That means, dragging in the Open tab will no more reorder the story-river while dragging, only at drop.

> IMO It adds unnecessary distraction to the river, if the user eg: "overshoots" to the top and then drags the tiddler back down by one or 2 elements.
> 
> If there is animation at the story river while dragging the Open tab elements is distracting.
> 
> My measurement for animation always is: **Does it serve a purpose**. If something only is animated, because we can, it should not be animated at all.
> 
> If an animation makes the user loose context or loose focus it should be not animated. So if I drag and drop elements in the open sidebar and the story river animates, **I do loose focus**

pmario commented on core/ui/SideBar/Open.tid:
```diff
@@ -14,7 +14,7 @@ caption: {{$:/language/SideBar/Open/Caption}}
 \whitespace trim
 
-\define content()
+\procedure content()
```
If you are at it, you can change all the single-quotes to double-quotes for string assignments

BurningTreeC replied:
Done @pmario

pmario commented:
LGTM - now

pmario commented on core/wiki/macros/dragndrop.tid:
```diff
@@ -151,6 +192,7 @@ code-body: yes
+				data-animate-hold="yes"
 				style.position="absolute"
 				style.top=<<tf.lr.get-before-top>>
 				style.left=<<tf.lr.get-before-left>>
```
Are these styles needed to be hardcoded, because JS code needs it that way. Or is that a development conveninece artefact? Same for line 212ff

BurningTreeC replied:
Hi @pmario , the `data-animate-hold` is needed actually for JS (the `animatelayout` widget)

pmario replied:
I meant `style.**`

BurningTreeC replied:
Hi @pmario, these styles here are functional, not aesthetic. They are needed like this so that the procedure works properly. Therefore I think it's correct to leave them here as `style.*` attributes

BurningTreeC commented:
Hi all, I've added Documentation and fixed all bugs (as far as I can tell)

The Docs contain an example for a simple Kanban board where we can drag&drop between lists. It works great, have a look. It's in the `make-live-reorderable-list Procedure` tiddler.

---

The core PageTemplate needed an eventcatcher added for the functionality to work.
Also, I had to add `$mode="inline"` to the transclusion in the core PageTemplate and the story-river Template to avoid `<p>` tags being added.

---

🥳 We can create Kanbans now!

BurningTreeC commented:
I marked this as ready for review. Tested on Chrome and Firefox.

BurningTreeC commented:
This is done @pmario @Jermolene 
Have a look at the features and the Kanban in the preview it's great I think.
I have no complaints anymore, everything seems to be solid and my issues are fixed.

pmario commented:
@BurningTreeC ... I am sorry. I wanted to create a PR to your fork in your repo, so you can review my suggestions.

But it seems my VSCode setup did directly push to your PR here in this rep. I am currently not sure how this happened. 

I did do every change in it's own commit, so it should be clear what's going on. The first ones are typo fixes, which should be no problem. 

But there are 2 code changes, [5f3d602](https://github.com/TiddlyWiki/TiddlyWiki5/pull/9507/commits/5f3d602ed18a844d49c93cb23b8c635970219c31) and [9547067](https://github.com/TiddlyWiki/TiddlyWiki5/pull/9507/commits/9547067e171543bb5a670d0bf2d7c93ff62ddc3e)

I do not mess with these changes any further, till I am sure what was configured wrong. 

The problem now is that you need to sync those changes back to your repo.

As I said. I am really sorry - I did not intend to directly mess with your PR.

------

Edit: I did update my VSCode configurations, that this should not happen again. 

<details><summary>Here is what exactly happened</summary>

I checked your branch out locally to review it. Because I created it by hand, VSCode's
Pull Requests sidebar did not associate it with this PR #9507, so I configured two things to make
the PR visible:

* `branch.btc-dragndrop.github-pr-owner-number` - this is the only one the VSCode PR sidebar needs
* `git branch --set-upstream-to=BTC/dragndrop` - **unnecessary, and the cause of this**

The second config wrote `branch.btc-dragndrop.remote = BTC` and `merge = refs/heads/dragndrop`,
which made your PR head branch the push destination for my local branch. Three things then
lined up:

1. My VSCode has `"git.confirmSync": false`, so Sync pulls and pushes with no dialog and no
   indication of where it is pushing.
2. I have `remote.pushDefault = origin` set, **which I assumed protected me**. It does not:
   VSCode pushes with an explicit remote, and an explicit remote overrides `pushDefault`.
3. This PR #9507 has "Allow edits by maintainers" enabled, which gives me write access to the head branch.
   So the push was accepted instead of rejected, which is what I had expected.

</details>

BurningTreeC commented:
Hi @pmario 😆 

It's totally fine, I've looked at your changes and they are totally Ok
